File: lab/backend/src/vm_lab/routers/rectify.py
"""POST /api/rectify — find + rectify each match into a canonical, model-frame crop
(`vm.ShapeMatch.model_frame_map`). See `crates/vision-metrology/src/matching/crop.rs`
for the Rust-side semantics this mirrors: `crop.rect` is in **model-frame coordinates**
(the same frame `ModelOut.roi` is expressed in), the crop's pixel size is fixed by the
spec alone, and `normalize_scale=True` renders at model scale regardless of the found
scale.

Rectified crops are cached in memory, keyed by `(image_id, model_id, index)` — this is a
single-user local workbench (`store.py`), so the cache always holds the most recent
`rectify` call's crops for that image/model pair, same spirit as `media.py`'s thumbnail
cache but not disk-persisted (purely derived, and specific to one request's crop spec).
`GET` the `crop_url` each match reports to fetch its PNG, `images/{id}/{tier}`-style
with a content-derived ETag.
"""

from __future__ import annotations

import hashlib
import io

from fastapi import APIRouter, HTTPException, Request, Response
from PIL import Image as PILImage

import vision_metrology as vm

from vm_lab.routers.find import run_find
from vm_lab.schemas import FindRequest, RectifyMatchOut, RectifyRequest, RectifyResponse
from vm_lab.store import store

router = APIRouter(prefix="/api/rectify", tags=["rectify"])

# (image_id, model_id, index) -> PNG bytes. In-memory only, like the thumbnail cache,
# and overwritten wholesale by each POST for that (image_id, model_id) pair.
_crop_cache: dict[tuple[str, str, int], bytes] = {}


def _crop_png(dst) -> bytes:  # noqa: ANN001 - a (H, W) uint8 ndarray from Map.apply_with_mask
    buf = io.BytesIO()
    PILImage.fromarray(dst).save(buf, "PNG")
    return buf.getvalue()


def _etag(payload: bytes) -> str:
    return f'"{hashlib.sha256(payload).hexdigest()[:16]}"'


@router.post("", response_model=RectifyResponse)
async def rectify(req: RectifyRequest) -> RectifyResponse:
    if req.image_id not in store.images:
        raise HTTPException(404, f"no such image: {req.image_id}")
    if req.model_id not in store.models:
        raise HTTPException(404, f"no such model: {req.model_id}")

    find_req = FindRequest(
        image_id=req.image_id,
        model_id=req.model_id,
        min_score=req.min_score,
        max_matches=req.max_matches,
    )
    matches = run_find(find_req)

    img = store.load_array(req.image_id)
    spec = vm.CropSpec(req.crop.rect, req.crop.px_per_unit, req.crop.normalize_scale)
    width, height = spec.output_size

    # Drop this pair's stale crops before caching the new set, so a shrinking
    # match count does not leave orphaned indices servable.
    for key in [k for k in _crop_cache if k[0] == req.image_id and k[1] == req.model_id]:
        del _crop_cache[key]

    out: list[RectifyMatchOut] = []
    for i, m in enumerate(matches):
        crop_map = m.model_frame_map(spec)
        dst, mask = crop_map.apply_with_mask(img, border_mode="constant")
        validity = float((mask == 255).mean())
        _crop_cache[(req.image_id, req.model_id, i)] = _crop_png(dst)
        out.append(
            RectifyMatchOut(
                index=i,
                x=m.x,
                y=m.y,
                angle=m.angle,
                scale=m.scale,
                score=m.score,
                support=m.support,
                level=m.level,
                validity=validity,
                crop_url=f"/api/rectify/{req.image_id}/{req.model_id}/{i}",
                width=width,
                height=height,
            )
        )
    return RectifyResponse(width=width, height=height, matches=out)


@router.get("/{image_id}/{model_id}/{index}")
async def get_crop(image_id: str, model_id: str, index: int, request: Request) -> Response:
    key = (image_id, model_id, index)
    payload = _crop_cache.get(key)
    if payload is None:
        raise HTTPException(
            404, "no cached crop for this image/model/index -- POST /api/rectify first"
        )
    tag = _etag(payload)
    if request.headers.get("if-none-match") == tag:
        return Response(status_code=304)
    return Response(content=payload, media_type="image/png", headers={"ETag": tag})

File: lab/frontend/src-tauri/src/commands/rectify.rs
//! `rectify`, `rectify_crop` — mirrors `lab/backend/src/vm_lab/routers/rectify.py`.
//!
//! No HTTP crop-cache URL here (there is no HTTP): `rectify` returns a `crop_key` per
//! match, cached in `AppState`, and `rectify_crop` returns the PNG bytes for one —
//! `tauriBackend.ts` turns that into an object URL, same as the browser backend's
//! `crop_url` becomes an `<img src>`.

use std::collections::HashMap;
use std::sync::Mutex;

use vision_metrology::matching::CropSpec;
use vision_metrology::warp::Interp;
use vm_primitives::{BorderMode, Rect2f};

use super::find::run_find;
use crate::error::{AppResult, not_found};
use crate::state::AppState;
use crate::types::{FindRequest, RectifyMatchOut, RectifyRequest, RectifyResponse};

/// `(image_id, model_id, index) -> PNG bytes`.
type CropKey = (String, String, usize);

/// In-memory only, overwritten wholesale by each `rectify` call for that
/// `(image_id, model_id)` pair — same spirit as the Python backend's own `_crop_cache`.
pub static CROP_CACHE: Mutex<Option<HashMap<CropKey, Vec<u8>>>> = Mutex::new(None);

fn crop_key(image_id: &str, model_id: &str, index: usize) -> String {
    format!("{image_id}/{model_id}/{index}")
}

pub fn rectify(state: &AppState, req: RectifyRequest) -> AppResult<RectifyResponse> {
    // `run_find` locks `state.images`/`state.models` itself, and so does
    // `AppState::decoded` below — neither may be called while this function
    // holds either lock (`std::sync::Mutex` is not reentrant, so the same
    // thread locking twice deadlocks rather than erroring).
    if !state
        .models
        .lock()
        .expect("models mutex poisoned")
        .contains_key(&req.model_id)
    {
        return Err(not_found("model", &req.model_id));
    }
    // The caller's whole search, not just its thresholds — see
    // `RectifyRequest`'s own docs for why anything less renumbers the matches.
    let find_req = FindRequest {
        image_id: req.image_id.clone(),
        model_id: req.model_id.clone(),
        min_score: req.min_score,
        max_matches: req.max_matches,
        angle_range: req.angle_range,
        scale_range: req.scale_range,
        refinement: req.refinement.clone(),
        min_contrast: req.min_contrast,
        tuning: req.tuning.clone(),
        roi: None,
    };
    let matches = run_find(state, &find_req)?;

    let image = state.decoded(&req.image_id)?;

    let spec = CropSpec {
        rect: Rect2f {
            x: req.crop.rect[0],
            y: req.crop.rect[1],
            width: req.crop.rect[2],
            height: req.crop.rect[3],
        },
        px_per_unit: req.crop.px_per_unit,
        normalize_scale: req.crop.normalize_scale,
    };
    let (width, height) = spec.output_size();

    let mut cache_guard = CROP_CACHE.lock().expect("crop cache mutex poisoned");
    let cache = cache_guard.get_or_insert_with(HashMap::new);
    cache.retain(|(iid, mid, _), _| !(iid == &req.image_id && mid == &req.model_id));

    let view = image.as_view();
    let mut out = Vec::with_capacity(matches.len());
    for (i, m) in matches.iter().enumerate() {
        let crop_map = m.model_frame_map(&spec);
        let mut dst = vec![0u8; width * height];
        let mut mask = vec![0u8; width * height];
        crop_map
            .apply_with_mask(
                &view,
                &mut dst,
                &mut mask,
                Interp::Bilinear,
                BorderMode::Constant(0),
            )
            .map_err(|e| crate::error::AppError(format!("rectify: {e}")))?;
        let validity = mask.iter().filter(|&&v| v == 255).count() as f32 / mask.len().max(1) as f32;

        let png = image::GrayImage::from_raw(width as u32, height as u32, dst)
            .ok_or_else(|| crate::error::AppError("could not build crop image".to_string()))?;
        let mut buf = Vec::new();
        png.write_to(&mut std::io::Cursor::new(&mut buf), image::ImageFormat::Png)
            .map_err(crate::error::AppError::from)?;
        cache.insert((req.image_id.clone(), req.model_id.clone(), i), buf);

        out.push(RectifyMatchOut {
            index: i,
            x: m.position.x,
            y: m.position.y,
            angle: m.angle(),
            scale: m.scale(),
            score: m.score,
            support: m.support,
            level: m.level,
            validity,
            crop_key: crop_key(&req.image_id, &req.model_id, i),
            width,
            height,
        });
    }

    Ok(RectifyResponse {
        width,
        height,
        matches: out,
    })
}

pub fn rectify_crop(image_id: &str, model_id: &str, index: usize) -> AppResult<Vec<u8>> {
    let mut cache_guard = CROP_CACHE.lock().expect("crop cache mutex poisoned");
    let cache = cache_guard.get_or_insert_with(HashMap::new);
    cache
        .get(&(image_id.to_string(), model_id.to_string(), index))
        .cloned()
        .ok_or_else(|| {
            crate::error::AppError(
                "no cached crop for this image/model/index -- call rectify first".to_string(),
            )
        })
}
